Source/GSOrderedSet.m

```diff
@@ -0,0 +1,407 @@
+#define GSI_ARRAY_TYPES       GSUNION_OBJ
+
+#define GSI_ARRAY_RELEASE(A, X)	[(X).obj release]
+#define GSI_ARRAY_RETAIN(A, X)	[(X).obj retain]
+
+#import "GNUstepBase/GSIArray.h"
+
+//  static SEL     memberSel;
+static SEL      privateCountOfSel;
+@interface GSOrderedSet : NSOrderedSet
+{
+@public
+  GSIArray_t array;
+}
+@end
+
+@interface GSMutableOrderedSet : NSMutableOrderedSet
+{
+@public
+  GSIArray_t array;
+@private
+  NSUInteger _version;
+}
+@end
+
+@interface GSOrderedSetEnumerator : NSEnumerator
+{
+  GSOrderedSet *set;
+  unsigned      current;
+  unsigned      count;
+}
+@end
+
+@implementation GSOrderedSetEnumerator
+- (id) initWithOrderedSet: (NSOrderedSet*)d
+{
+  self = [super init];
+  if (self != nil)
+    {
+      set = (GSOrderedSet*)RETAIN(d);
+      current = 0;
+      count = GSIArrayCount(&set->array);
+    }
+  return self;
+}
+
+- (id) nextObject
+{
+  if(current < count)
+    {
+      GSIArrayItem item = GSIArrayItemAtIndex(&set->array, current);
+      current++;
+      return (id)(item.obj);
+    }
+  return nil;
+}
+
+- (void) dealloc
+{
+  RELEASE(set);
+  [super dealloc];
+}
+@end
+
+@implementation GSOrderedSet
+
+static Class	setClass;
+static Class	mutableSetClass;
+
++ (void) initialize
+{
+  if (self == [GSOrderedSet class])
+    {
+      setClass = [GSOrderedSet class];
+      mutableSetClass = [GSMutableOrderedSet class];
+      privateCountOfSel = @selector(_countForObject:);
+    }
+}
+
+- (id) copyWithZone: (NSZone*)z
+{
+  return RETAIN(self);
+}
+
+- (void) dealloc
+{
+  GSIArrayEmpty(&array);
+  [super dealloc];
+}
+
+- (NSUInteger) hash
+{
+  return [self count];
+}
+
+- (id) init
+{
+  return [self initWithObjects: 0 count: 0];
+}
+
+- (NSEnumerator*) objectEnumerator
+{
+  return AUTORELEASE([[GSOrderedSetEnumerator alloc] initWithOrderedSet: self]);
+}
+
+- (NSUInteger) countByEnumeratingWithState: (NSFastEnumerationState*)state
+                                   objects: (id*)stackbuf
+                                     count: (NSUInteger)len
+{
+  //state->mutationsPtr = (unsigned long *)self;
+  //return GSIMapCountByEnumeratingWithStateObjectsCount
+  //            (&map, state, stackbuf, len);
+
+  return [self count];
+}
+
+- (NSUInteger) sizeInBytesExcluding: (NSHashTable*)exclude
+{
+  NSUInteger	size = GSPrivateMemorySize(self, exclude);
+
+  if (size > 0)
+    {
+      NSUInteger count = [self count];
+      NSUInteger i = 0;
+
+      for(i = 0; i < count; i++)
+	{
+	  GSIArrayItem item = GSIArrayItemAtIndex(&array, i);
+          size += [item.obj sizeInBytesExcluding: exclude];
+        }
+    }
+  return size;
+}
+
+// Put required overrides here...
+- (NSUInteger) count
+{
+  return GSIArrayCount(&array);
+}
+
+- (id) objectAtIndex: (NSUInteger)index
+{
+  GSIArrayItem item = GSIArrayItemAtIndex(&array, index);
+  return item.obj;
+}
+
+- (id) objectAtIndexedSubscript: (NSUInteger)index
+{
+  return[self objectAtIndex: index];
+}
+
+- (BOOL) containsObject: (id)anObject
```

Comment:
This method could be moved to the abstract class.

Reply:
Or rather not. We should make this method really fast by using a hashing structure in addition to the array. This is the main difference between this class and NSArray, so this may actually be the most important part of the class.

Reply:
Okay, cool.

Reply:
Since implementing the hashing structure is more involved I'm going to submit a change request for it.